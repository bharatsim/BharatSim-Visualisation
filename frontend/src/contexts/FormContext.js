import React, { useContext } from 'react';

const FormContext = React.createContext({});

const FormProvider = FormContext.Provider;

const useFormContext = () => useContext(FormContext);

export { FormContext, FormProvider, useFormContext };
