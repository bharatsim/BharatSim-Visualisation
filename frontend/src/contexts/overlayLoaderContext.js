import React from 'react';

const overlayLoaderContext = React.createContext({
  stopLoader: null,
  startLoader: null,
  showError: null,
  isLoading: false,
});

const OverlayLoaderContextProvider = overlayLoaderContext.Provider;

export { overlayLoaderContext, OverlayLoaderContextProvider };
