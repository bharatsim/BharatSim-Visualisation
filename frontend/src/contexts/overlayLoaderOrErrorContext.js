import React from 'react';

const overlayLoaderOrErrorContext = React.createContext({
  stopLoader: null,
  startLoader: null,
  showError: null,
  isLoading: false,
});

const OverlayLoaderOrErrorContextProvider = overlayLoaderOrErrorContext.Provider;

export { overlayLoaderOrErrorContext, OverlayLoaderOrErrorContextProvider };
