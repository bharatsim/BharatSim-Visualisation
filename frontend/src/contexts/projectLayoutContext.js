import React from 'react';

const projectLayoutContext = React.createContext({
  projectMetadata: {
    id: undefined,
    name: undefined,
  },
  selectedDashboardMetadata: {
    name: undefined,
    _id: undefined,
  },
  addDashboard: null,
  deleteDashboard: null,
});

const ProjectLayoutProvider = projectLayoutContext.Provider;

export { projectLayoutContext, ProjectLayoutProvider };
