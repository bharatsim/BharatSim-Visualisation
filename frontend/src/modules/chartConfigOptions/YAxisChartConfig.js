import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Box, fade, makeStyles, Typography } from '@material-ui/core';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import plusIcon from '../../assets/images/plus.svg';
import deleteIcon from '../../assets/images/delete.svg';
import Dropdown from '../../uiComponent/Dropdown';
import { convertObjectArrayToOptionStructure } from '../../utils/helper';

const useStyles = makeStyles((theme) => {
  return {
    fieldContainer: {
      display: 'flex',
      justifyContent: 'space-between',
      padding: theme.spacing(2),
      borderColor: fade(theme.colors.primaryColorScale['500'], 0.24),
      border: '1px solid',
      borderRadius: theme.spacing(1),
      marginBottom: theme.spacing(4),
      '&:last-child': {
        marginBottom: 0,
      },
    },
    addMetricButtonContainer: {
      display: 'flex',
      justifyContent: 'flex-end',
      marginTop: theme.spacing(4),
    },
  };
});

function YAxisChartConfig({ headers, updateConfigState, configKey, error }) {
  const [selectedValues, setSelectedValues] = useState({ 'dropdown-y-0': '' });
  const [axisCount, setAxisCount] = useState(1);
  const classes = useStyles();

  function handleYChange(id, selectedValue) {
    setSelectedValues((prevState) => {
      const newState = { ...prevState, [id]: selectedValue };
      updateConfigState(configKey, Object.values(newState));
      return newState;
    });
  }

  function addAxisField() {
    setSelectedValues((prevState) => ({ ...prevState, [`dropdown-y-${axisCount}`]: '' }));
    setAxisCount((prevCount) => prevCount + 1);
  }
  function deleteAxisField(fieldId) {
    setSelectedValues((prevState) => {
      const newState = { ...prevState };
      delete newState[fieldId];
      updateConfigState(configKey, Object.values(newState));
      return newState;
    });
  }

  return (
    <>
      <Box mb={1} pl={2}>
        <Typography variant="subtitle2"> Y-axis</Typography>
      </Box>
      {Object.keys(selectedValues).map((key) => (
        <Box className={classes.fieldContainer} key={key}>
          <Dropdown
            options={convertObjectArrayToOptionStructure(headers, 'name')}
            onChange={(selectedValue) => handleYChange(key, selectedValue)}
            id={key}
            label="select y axis"
            error={error}
            value={selectedValues[key] || ''}
          />
          <IconButton onClick={() => deleteAxisField(key)} data-testid={`delete-button-${key}`}>
            <img src={deleteIcon} alt="icon" />
          </IconButton>
        </Box>
      ))}
      <Box className={classes.addMetricButtonContainer}>
        <Button
          variant="contained"
          color="secondary"
          size="small"
          onClick={addAxisField}
          startIcon={<img src={plusIcon} alt="icon" />}
        >
          Add Metric
        </Button>
      </Box>
    </>
  );
}

YAxisChartConfig.defaultProps = {
  error: '',
};

YAxisChartConfig.propTypes = {
  headers: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
    }),
  ).isRequired,
  updateConfigState: PropTypes.func.isRequired,
  configKey: PropTypes.string.isRequired,
  error: PropTypes.string,
};

export default YAxisChartConfig;
