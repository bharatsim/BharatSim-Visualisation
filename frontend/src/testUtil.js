// eslint-disable-next-line import/no-extraneous-dependencies
import { fireEvent, within } from '@testing-library/react';

export function selectDropDownOption(container, dropDownId, optionId) {
  const dropDown = container.getByTestId(dropDownId);
  fireEvent.mouseDown(within(dropDown).getByRole('button'));
  const options = within(
    within(document.getElementById(`menu-${dropDownId}`)).getByRole('listbox'),
  );
  fireEvent.click(options.getByTestId(`${dropDownId}-${optionId}`));
}
