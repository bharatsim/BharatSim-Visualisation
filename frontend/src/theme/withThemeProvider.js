import React from 'react';
import { MuiThemeProvider, CssBaseline } from '@material-ui/core';

import theme from './theme';

function withThemeProvider(WrappedComponent) {
  return function WithTheme(props) {
    return (
      <MuiThemeProvider theme={theme}>
        <CssBaseline />
        <WrappedComponent {...props} />
      </MuiThemeProvider>
    );
  };
}

export default withThemeProvider;
