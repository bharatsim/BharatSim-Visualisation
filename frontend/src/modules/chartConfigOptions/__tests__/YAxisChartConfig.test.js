import React from 'react';
import { render } from '@testing-library/react';
import { fireEvent } from '@testing-library/dom';
import { Form } from 'react-final-form';
import arrayMutators from 'final-form-arrays';

import { selectDropDownOption } from '../../../testUtil';
import withThemeProvider from '../../../theme/withThemeProvider';
import YAxisChartConfig from '../YAxisChartConfig';

import { FormProvider } from '../../../contexts/FormContext';

const headers = [
  { name: 'a', type: 'number' },
  { name: 'b', type: 'number' },
  { name: 'c', type: 'number' },
];

const TestForm = ({ onSubmit, isEditMode }) => {
  const configKey = 'yAxis';
  const initialValues = {
    yAxis: [{ name: 'b' }, { name: 'c' }],
  };
  return (
    <Form
      onSubmit={onSubmit}
      initialValues={isEditMode ? initialValues : undefined}
      mutators={{ ...arrayMutators }}
      render={({ handleSubmit }) => (
        <FormProvider
          value={{
            isEditMode: !!isEditMode,
            registerDatasource: jest.fn(),
            unRegisterDatasource: jest.fn(),
          }}
        >
          <form onSubmit={handleSubmit}>
            <YAxisChartConfig configKey={configKey} headers={headers} isEditMode={isEditMode} />
            <button type="submit">submit</button>
          </form>
        </FormProvider>
      )}
    />
  );
};

describe('<YAxisConfig />', () => {
  const FormForYAxisConfig = withThemeProvider(TestForm);

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should not add extra dropbox if form already has one or more yaxis field', async () => {
    const onSubmit = jest.fn();
    const renderedContainer = render(<FormForYAxisConfig onSubmit={onSubmit} isEditMode />);
    const { getAllByText } = renderedContainer;
    const yAxisDropdown = getAllByText('select y axis');

    expect(yAxisDropdown.length).toBe(2);
  });

  it('should add dropbox if form has zero yaxis field', async () => {
    const onSubmit = jest.fn();
    const renderedContainer = render(<FormForYAxisConfig onSubmit={onSubmit} />);
    const { getAllByText } = renderedContainer;
    const yAxisDropdown = getAllByText('select y axis');

    expect(yAxisDropdown.length).toBe(1);
  });

  it('should call on submit with y axis filedArray config', async () => {
    const onSubmit = jest.fn();
    const renderedContainer = render(<FormForYAxisConfig onSubmit={onSubmit} />);
    const { getByText } = renderedContainer;

    const addFieldButton = getByText('Add Metric');
    fireEvent.click(addFieldButton);

    selectDropDownOption(renderedContainer, 'y-axis-dropdown-0', 'a');
    selectDropDownOption(renderedContainer, 'y-axis-dropdown-1', 'b');

    fireEvent.click(renderedContainer.getByText('submit'));

    expect(onSubmit).toHaveBeenCalledWith(
      {
        yAxis: [{ name: 'a' }, { name: 'b' }],
      },
      expect.anything(),
      expect.anything(),
    );
  });

  it('should add y axis field on click of add metric button', () => {
    const onSubmit = jest.fn();
    const { getByText, queryByTestId } = render(<FormForYAxisConfig onSubmit={onSubmit} />);

    const addFieldButton = getByText('Add Metric');

    fireEvent.click(addFieldButton);

    const newField = queryByTestId('y-axis-dropdown-1');

    expect(newField).toBeInTheDocument();
  });

  it('should delete y axis field on click of delete button', () => {
    const onSubmit = jest.fn();
    const { getByText, queryByTestId, getByTestId } = render(
      <FormForYAxisConfig onSubmit={onSubmit} />,
    );

    const addFieldButton = getByText('Add Metric');

    fireEvent.click(addFieldButton);

    const newField = queryByTestId('y-axis-dropdown-1');

    expect(newField).toBeInTheDocument();

    const deleteButton = getByTestId('delete-button-1');
    fireEvent.click(deleteButton);

    expect(newField).not.toBeInTheDocument();
  });

  it('should not have delete button if only one yaxis config is present', () => {
    const onSubmit = jest.fn();
    const renderedContainer = render(<FormForYAxisConfig onSubmit={onSubmit} />);
    const { queryAllByAltText } = renderedContainer;

    const deleteIcons = queryAllByAltText('delete-icon');

    expect(deleteIcons.length).toBe(0);
  });
});
