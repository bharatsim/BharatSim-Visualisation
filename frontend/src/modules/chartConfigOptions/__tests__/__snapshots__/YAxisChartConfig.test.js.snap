// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<YAxisChartConfig /> should match snapshot 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-3"
  >
    <h6
      class="MuiTypography-root MuiTypography-subtitle2"
    >
       Y-axis
    </h6>
  </div>
  <div
    class="MuiBox-root MuiBox-root-4 makeStyles-fieldContainer-1"
  >
    <div
      class="MuiFormControl-root makeStyles-formControl-5"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
        data-shrink="false"
        id="dropdown-label"
      >
        select y axis
      </label>
      <div
        class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
        data-testid="dropdown-y-0"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="dropdown-label dropdown-y-0"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-filled MuiInputBase-input MuiFilledInput-input"
          id="dropdown-y-0"
          role="button"
          tabindex="0"
        >
          <span>
            â€‹
          </span>
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput"
          tabindex="-1"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconFilled"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiIconButton-root"
      data-testid="delete-button-dropdown-y-0"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <img
          alt="icon"
          src="delete.svg"
        />
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    class="MuiBox-root MuiBox-root-7 makeStyles-addMetricButtonContainer-2"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-containedSizeSmall MuiButton-sizeSmall"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeSmall"
        >
          <img
            alt="icon"
            src="plus.svg"
          />
        </span>
        Add Metric
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
